cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,378 @@
+      if (not target[i]->nullable()) {
+        // Target must have a null mask if the source is not valid
+        auto mask = create_null_mask(target[i]->size(), mask_state::ALL_VALID, stream, mr);
+        target[i]->set_null_mask(std::move(mask), 0);
+      }
+
+      auto target_view = mutable_column_device_view::create(
+        target[i]->mutable_view(), stream);
+
+      auto bitmask_kernel = source_is_valid
+        ? marking_bitmask_kernel<true, decltype(scatter_map)>
+        : marking_bitmask_kernel<false, decltype(scatter_map)>;
+      bitmask_kernel<<<grid_size, block_size, 0, stream>>>(
+        *target_view, scatter_map, num_scatter_rows);
+    }
+  }
+}
+
+template <typename MapIterator>
+struct column_scalar_scatterer {
+  template <typename T, std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(std::unique_ptr<scalar> const& source,
+      MapIterator scatter_iter, size_type scatter_rows, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto result = std::make_unique<column>(target, stream, mr);
+    auto result_view = result->mutable_view();
+
+    // Use permutation iterator with constant index to dereference scalar data
+    auto scalar_impl = static_cast<scalar_type_t<T>*>(source.get());
+    auto scalar_iter = thrust::make_permutation_iterator(
+      scalar_impl->data(), thrust::make_constant_iterator(0));
+
+    thrust::scatter(rmm::exec_policy(stream)->on(stream), scalar_iter,
+      scalar_iter + scatter_rows, scatter_iter,
+      result_view.begin<T>());
+
+    return result;
+  }
+
+  template <typename T, std::enable_if_t<not is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(std::unique_ptr<scalar> const& source,
+      MapIterator scatter_iter, size_type scatter_rows, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto const scalar_impl = static_cast<string_scalar*>(source.get());
+    auto const source_view = string_view(scalar_impl->data(), scalar_impl->size());
+    auto const begin = thrust::make_constant_iterator(source_view);
+    auto const end = begin + scatter_rows;
+    return strings::detail::scatter(begin, end, scatter_iter, target, mr, stream);
+  }
+};
+
+struct scatter_scalar_impl {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value
+      and not std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      std::vector<std::unique_ptr<scalar>> const& source,
+      column_view const& indices, table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    if (check_bounds) {
+      auto const begin = -target.num_rows();
+      auto const end = target.num_rows();
+      auto bounds = bounds_checker<T>{begin, end};
+      CUDF_EXPECTS(thrust::all_of(rmm::exec_policy(stream)->on(stream),
```

Comment:
Don't use `thrust::all_of`, it's slow: https://github.com/thrust/thrust/issues/1016

Reply:
Shoot, that explains why the old code I ported from was using count_if...

Reply:
I'll replace `all_of` with `count_if` in https://github.com/rapidsai/cudf/pull/3448